core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java

```diff
@@ -1201,22 +1202,19 @@ private RecordCursorFactory generateLimit(RecordCursorFactory factory, QueryMode
         ExpressionNode limitLo = model.getLimitLo();
         ExpressionNode limitHi = model.getLimitHi();
 
-        if (limitLo == null && limitHi == null) {
+        if ((limitLo == null && limitHi == null) || factory.implementsLimit()) {
             return factory;
         }
 
-        final Function loFunc;
-        final Function hiFunc;
+        final Function loFunc = getLoFunction(executionContext, limitLo);
+        final Function hiFunc = getHiFunction(executionContext, limitHi);
 
-        if (limitLo == null) {
-            loFunc = LongConstant.ZERO;
-        } else {
-            loFunc = functionParser.parseFunction(limitLo, EmptyRecordMetadata.INSTANCE, executionContext);
-            final int type = loFunc.getType();
-            if (limitTypes.excludes(type)) {
-                throw SqlException.$(limitLo.position, "invalid type: ").put(ColumnType.nameOf(type));
-            }
-        }
+        return new LimitRecordCursorFactory(factory, loFunc, hiFunc);
+    }
+
+    @Nullable
+    private Function getHiFunction(SqlExecutionContext executionContext, ExpressionNode limitHi) throws SqlException {
```

Comment:
I reworked both functions to use the same internal function . I didn't get rid of getHiFunction or getLoFunction because there are at least two call sites and inlining method would expose default lo value (that'd be fragile) .